kernel/src/debug.rs

```diff
@@ -289,20 +325,39 @@ macro_rules! debug_gpio {
 ///////////////////////////////////////////////////////////////////
 // debug! and debug_verbose! support
 
+/// A trait for writing debug output.
+///
+/// This can be used for example to implement asynchronous or synchronous
+/// writers, and buffered or unbuffered writers. Various platforms may have
+/// in-memory logs, memory mapped "endless" FIFOs, JTAG support for output,
+/// etc.
+pub trait DebugWriter {
+    /// Write bytes to output with overflow notification.
+    ///
+    /// The `overflow` slice is used as a message to be appended to the end of
+    /// the available buffer if it becomes full.
+    fn write(&self, bytes: &[u8], overflow_message: &[u8]) -> usize;
+
+    /// Available length of the internal buffer if limited.
+    ///
+    /// If the buffer can support a write of any size, this returns `None`.
+    fn available_len(&self) -> Option<usize>;
+
+    /// How many bytes are buffered and not yet written.
+    fn to_write_len(&self) -> usize;
+
+    /// Publish bytes from the internal buffer to the output.
+    ///
+    /// Returns how many bytes were written.
+    fn publish(&self) -> usize;
+
+    /// Flush any buffered bytes to the output writer.
+    fn flush(&self, writer: &mut dyn IoWrite);
+}
```

Comment:
This should document the difference between `publish` and `flush`, and when to use which.

Reply:
Tried to clarify this.